### ISSUE TITLE
Bizarre error with variable _C2 in VS2008

### ISSUE THREAD

**Issue by thesamovar** [brian-team/brian2#472]
I just had a peculiar bug where if there is a variable named ``C2`` then Weave with MSVS2008 crashes complaining about the left-hand side not being an l-value on a line where the LHS is ``_C2``. I assume that some header file defines ``_C2`` to be something. I don't think there's much we can do about this? Or is there a general way we could handle it? Perhaps we could undef all variables and then restore them afterwards, e.g.

```
#ifdef _C2
#define _BRIAN2_TEMP_VAR_STORAGE__C2 _C2
#undef _C2
#endif
... (Brian code here)
#ifdef _BRIAN2_TEMP_VAR_STORAGE__C2
#define _C2 _BRIAN2_TEMP_VAR_STORAGE__C2
#undef _BRIAN2_TEMP_VAR_STORAGE__C2
#endif
```

And those blocks would be repeated for each variable name.

Worth it?

**Comment by mstimberg** [comment]
Just to make sure: it is a general weave issue, not something that we introduce? What I mean, does something like `weave.inline('double _C2 = 1.0;')` fail with the same error?

I guess in your example `C2` is a state variable and `_C2` is the intermediate value generated during the state update? I think messing with the defines is a bit of overkill in that case, but instead of using `_{name}` we should probably rather use something like `_brian_{name}` to avoid potential problems of that sort.

**Comment by thesamovar** [comment]
I just tried that inline on my computer and it worked fine, but actually the problematic code was on someone else's machine. I'll get him to send me a copy and see if it was something specific to his installation or not.

Your one works fine but not if the user specifies their own variable that clashes with something. The undef method is maybe guaranteed to work no matter what variables have been defined.

**Comment by mstimberg** [comment]
> Your one works fine but not if the user specifies their own variable that clashes with something.

Right, but I don't think any standard header file will bring in macros that don't start with an underscore, use some specific prefix/suffix, etc., does it? If we want to check macro names, then I think we should raise an error if we encounter a clash, i.e. for each variable/function name
```C++
#ifdef varname
#error There's a macro definition clashing with the variable name "varname"
#endif
```
Or simply unset every clashing macro...

I think your approach of restoring the macro afterwards does not work like this. I tried the following code
```C++
// This would produce the error of an invalid LHS if assigned to it
#define _C2 1.0

#ifdef _C2
#define _BRIAN2_TEMP_VAR_STORAGE__C2 _C2
#undef _C2
#endif

// Brian code here, using _C2
double _C2;
_C2 = 3.0;

#ifdef _BRIAN2_TEMP_VAR_STORAGE__C2
#define _C2 _BRIAN2_TEMP_VAR_STORAGE__C2
#undef _BRIAN2_TEMP_VAR_STORAGE__C2
#endif

// Non-Brian code, using _C2
_C2
```

And the pre-processor expands it to:
```C++
double _C2;
_C2 = 3.0;

_BRIAN2_TEMP_VAR_STORAGE__C2
```

**Comment by thesamovar** [comment]
Ah that's a shame, I thought that undef thing worked but apparently not. Yes, maybe your ifdef/error technique is better then: at least that way the user knows what the problem is rather than getting a very obscure error message about l-values.

**Comment by thesamovar** [comment]
This script reproduces the error on my computer:

```
from brian2 import *

eqs = '''
dC2/dt = 0/second : 1
'''
G = NeuronGroup(1, eqs)
run(1*ms)
```

**Comment by thesamovar** [comment]
With Cython it works fine, so it's something that weave does that Cython doesn't do.

**Comment by mstimberg** [comment]
And just to confirm that it is MSVC-specific: the example works fine on my Linux machine.

**Comment by thesamovar** [comment]
Well I found it anyway, it's in a header ``yvals.h`` and has value 1. 

```
		/* INTEGER PROPERTIES */
#define _C2			1	/* 0 if not 2's complement */
```
